test-d/set-required-deep.ts

```diff
@@ -0,0 +1,22 @@
+import {expectType} from 'tsd';
+import type {SetRequiredDeep} from '../index';
+
+// Update an optional nested key to required.
+declare const variation1: SetRequiredDeep<{a?: number; b?: {c?: string}}, 'b.c'>;
+expectType<{a?: number; b: {c: string}}>(variation1);
+
+// Update a root key to required.
+declare const variation2: SetRequiredDeep<{a?: number; b?: {c?: string}}, 'a'>;
+expectType<{a: number; b?: {c?: string}}>(variation2);
+
+// Keeps required key as required
+declare const variation3: SetRequiredDeep<{a: number; b?: {c?: string}}, 'a'>;
+expectType<{a: number; b?: {c?: string}}>(variation3);
+
+// Update optional keys to required in a union.
+declare const variation4: SetRequiredDeep<{a?: '1'; b?: {c?: boolean}} | {a?: '2'; b?: {c?: boolean}}, 'a' | 'b.c'>;
+expectType<{a: '1'; b: {c: boolean}} | {a: '2'; b: {c: boolean}}>(variation4);
+
```

Comment:
add tests:
```
SetRequiredDeep<{a?: number; array?: Array<{b?: number}>}, 'array.${number}.b'>;

SetRequiredDeep<{a?: number; array?: Array<{b?: number}>}, 'array[0].b'>;

SetRequiredDeep<{a?: number; array?: [{b?: number}, {c?: number}]}, 'array[1].c'>;
```

Reply:
It's not working for array nested types, because the combination of the required and not required keys do not merge the array back together after setting keys to required.

So the test shows this error, which is legit:
<img width="908" alt="Screenshot 2024-08-19 at 11 53 41" src="https://github.com/user-attachments/assets/c49fd4e3-498b-49af-8e20-5a8262043156">
<img width="1240" alt="Screenshot 2024-08-19 at 11 54 38" src="https://github.com/user-attachments/assets/3ab144e7-1863-4491-b911-b3f86a4fed49">

Is there a type in **type-fest** that I could use to handle the merging of nested array types properly ? I searched but couldn't find one.

Thanks for your help!

Reply:
`&` is not effect the final result, I think this is a little bug on `SimplifyDeep`, you can test by `expectAssignable`:
```
expectAssignable<SetRequiredDeep<{a?: number; array?: Array<{b?: number; c?: string}>}, `array.${number}.b`>>({array: [{b: 1}]});
```

Reply:
Actually, the problem with my current type definition of `SetRequiredDeep` is that : 
If I write 
```typescript
type Result = SetRequiredDeep<{ array?: {b?: number; c?: string}[] }, `array.${number}.b`>
```

I want the result to be :
```typescript
type Result = { array?: {b: number; c?: string}[] }
```

But for now the result will be :
```typescript
type Result = { array: {b: number}[] &  {c?: string}[] }
```
which is incorrect...

Reply:
I guess I need to define the `SetRequiredDeep` type recursively right? So I can't use `RequiredDeep` util in my definition

Reply:
> Actually, the problem with my current type definition of `SetRequiredDeep` is that : If I write
> 
> ```ts
> type Result = SetRequiredDeep<{ array?: {b?: number; c?: string}[] }, `array.${number}.b`>
> ```

If your query is about the optionality of the "array", I think `array` is correct, not `array?`.

If about `{b: number; c?: string}[]` and `{b: number}[] &  {c?: string}[]`, [this answer](https://stackoverflow.com/questions/58088420/any-behavior-difference-between-a-number-b-string-and-a-number-b) can help you. (I think this should be handled by `SimplifyDeep` in your code, but it's not being done now)

Reply:
Thanks for the explanation!

I added the 3 tests you asked for but since SimplifyDeep does not handle the merging of array types, the tests fail (as you can see in the CI). Should I comment them for now?

Reply:
> `&` is not effect the final result, `SimplifyDeep` only make result more beautiful and not change the result. I think this is a little bug on `SimplifyDeep`, you can test by `expectAssignable`:
> 
> ```ts
> expectAssignable<SetRequiredDeep<{a?: number; array?: Array<{b?: number; c?: string}>}, `array.${number}.b`>>({array: [{b: 1}]});
> ```

use this instead 👆🏻

Reply:
🙏